pulser/_json_coders.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2020 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import importlib
+from json import JSONEncoder, JSONDecoder
+
+import numpy as np
+
+from pulser.utils import obj_to_dict
+from pulser.parametrized import Variable
+
+
+class PulserEncoder(JSONEncoder):
+    def default(self, o):
+        if hasattr(o, "_to_dict"):
+            return o._to_dict()
+        elif type(o) == type:
+            return obj_to_dict(o, _build=False, _name=o.__name__)
+        elif isinstance(o, np.ndarray):
+            return obj_to_dict(o, o.tolist(), _name="array")
+        elif isinstance(o, set):
+            return obj_to_dict(o, list(o))
+        else:
+            return JSONEncoder.default(self, o)
+
+
+class PulserDecoder(JSONDecoder):
+    def __init__(self, *args, **kwargs):
+        self.vars = {}
+        super().__init__(object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, obj):
+        try:
+            build = obj["_build"]
+            obj_name = obj["__name__"]
+            module_str = obj["__module__"]
+        except KeyError:
+            return obj
+
+        if (obj_name == "Variable" and
+           module_str == "pulser.parametrized.variable"):
+            var_name = obj["name"]
+            try:
+                var = self.vars[var_name]
+                assert var.name == var_name
+                assert var.dtype == obj["dtype"]
+                assert var.size == obj["size"]
```

Comment:
The error raised by assert is not very explicit if I remember correctly. Might be worth having a more detailed error message.
edit: I seems you can add a custom message to assert :)
```python
assert var.dtype == obj["dtype"], f"variable {var_name} already defined with different parameters"
```

Reply:
Cool! I didn't know this either, I will use it. I used `assert` statements more as a safety check that anything else, it seemed excessive to be raising a bunch of special errors. I agree the lack of an error message is a downside, but this solves that!